stdlib/http/src/main/java/org/ballerinalang/net/http/websocket/WebSocketUtil.java

```diff
@@ -169,62 +180,199 @@ public static String getErrorMessage(Throwable err) {
+     * Check whether the client's config has the `retryConfig` property or not.
+     *
+     * @param webSocketClient - the WebSocket client.
+     * @return If the client's config has the retry config, then return true.
+     */
+    public static boolean hasRetryConfig(ObjectValue webSocketClient) {
+        return webSocketClient.getMapValue(CLIENT_ENDPOINT_CONFIG).
+                getMapValue(WebSocketConstants.RETRY_CONFIG) != null;
+    }
+
+    private static void waitForHandshake(CountDownLatch countDownLatch) {
+        try {
+            // Wait for 5 minutes before the timeout.
+            if (!countDownLatch.await(5, TimeUnit.MINUTES)) {
+                throw new WebSocketException(WebSocketConstants.ErrorCode.WsGenericError,
+                        "Waiting for WebSocket handshake has not been successful", WebSocketUtil.createErrorCause(
+                        "Connection timeout", IOConstants.ErrorCode.ConnectionTimedOut.errorCode(),
+                        IOConstants.IO_PACKAGE_ID));
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new WebSocketException(ERROR_MESSAGE + e.getMessage());
+        }
+    }
+
+    /**
+     * Establish the connection with the endpoint and set the listener into it.
+     *
+     * @param webSocketClient - the WebSocket client.
+     * @param wsService - the WebSocket service.
+     */
+    public static void establishWebSocketConnection(ObjectValue webSocketClient, WebSocketService wsService) {
+        WebSocketClientConnectorListener clientConnectorListener = (WebSocketClientConnectorListener) webSocketClient.
+                getNativeData(WebSocketConstants.CLIENT_LISTENER);
+        WebSocketClientConnector clientConnector = (WebSocketClientConnector) webSocketClient.
+                getNativeData(WebSocketConstants.CLIENT_CONNECTOR);
+        boolean readyOnConnect = webSocketClient.getMapValue(CLIENT_ENDPOINT_CONFIG).getBooleanValue(
+                WebSocketConstants.CLIENT_READY_ON_CONNECT);
+        ClientHandshakeFuture handshakeFuture = clientConnector.connect();
+        handshakeFuture.setWebSocketConnectorListener(clientConnectorListener);
+        if (webSocketClient.getNativeData(WebSocketConstants.RETRY_CONFIG) != null &&
+                ((RetryContext) webSocketClient.getNativeData(WebSocketConstants.RETRY_CONFIG)).
+                        isFirstConnectionMadeSuccessfully()) {
+            CountDownLatch countDownLatch = new CountDownLatch(1);
+            handshakeFuture.setClientHandshakeListener(new WebSocketClientHandshakeListener(webSocketClient, wsService,
+                    clientConnectorListener, readyOnConnect, countDownLatch));
+            waitForHandshake(countDownLatch);
+        } else {
+            handshakeFuture.setClientHandshakeListener(new WebSocketClientHandshakeListener(webSocketClient, wsService,
+                    clientConnectorListener, readyOnConnect, null));
```

Comment:
Why pass null here?

Reply:
This countdown latch is to wait for success of the handshake and fail the connection. If you want a countdown latch to hold the thread use a different one. I think I mentioned clearly that it is wrong to pass null here.